RandBLAS/trig_skops.hh

```diff
@@ -316,125 +298,157 @@ namespace RandBLAS {
-            fht_left_row_major(buff, log_n, num_rows, num_cols);
+            fht_left_row_major(A, log_n, num_rows, num_cols);
         else if(!left && layout == blas::Layout::ColMajor)
-            fht_right_col_major(buff, log_n, num_rows, num_cols);
+            fht_right_col_major(A, log_n, num_rows, num_cols);
         else
-            fht_right_row_major(buff, log_n, num_rows, num_cols);
+            fht_right_row_major(A, log_n, num_rows, num_cols);
     }
 }
 
 
 namespace RandBLAS::trig {
+template <SignedInteger sint_t = int64_t>
+struct HadamardMixingOp{
+    sint_t* diag_scale;
+    sint_t* selected_idxs;
+    blas::Layout layout;
+    int64_t m;
+    int64_t n;
+    int64_t d;
+    bool left;
+    bool filled = false; // will be updated by `miget`
+
+    // Constructor
+    HadamardMixingOp(bool left,
+                     blas::Layout layout,
+                     int64_t m,
+                     int64_t n,
+                     int64_t d
+                    ) : left(left), layout(layout), m(m), n(n), d(d) {
+        if(left)
+            diag_scale = new sint_t[m];
+        else
+            diag_scale = new sint_t[n];
+        selected_idxs = new sint_t[d];
+    }
+
+    // Destructor
+    ~HadamardMixingOp() {
+        free(this->diag_scale);
+        free(this->selected_idxs);
+    }
+
+    private:
+};
+
+/*
+ * A free-function that performs an inversion of a matrix transformed by `lmiget | rmiget`
+ * the inversion is also performed in-place
+*/
+template <typename T, SignedInteger sint_t = int64_t>
+void invert(
+    HadamardMixingOp<sint_t> &hmo, // details about the transform
+    T* SA // sketched matrix
+) {
+    // We have to make sure we apply the operation in the inverted order
+    // with the operations appropriately inverted
+    randblas_error_if_msg(!hmo.filled, "You are trying to call `invert` on an untransformed matrix,\
+                                       please call `miget` on your matrix before calling `invert`");
+
+    // Creating a vector out of `selected_idxs` to be able to conveniently reverse
+    std::vector<sint_t> selected_idxs(hmo.selected_idxs, hmo.selected_idxs + hmo.d);
+    // Reversing the indices for performing the inverse
+    std::reverse(selected_idxs.begin(), selected_idxs.end());
+
+    //Step 1: Permute the rows/cols
+        // Perform the permutation (for the way we define permutations
+        // invSelected_idxs = reverse(Selected_idxs))
+    if(hmo.left)
+        permute_rows_to_top(hmo.layout, hmo.m, hmo.n, selected_idxs.data(), hmo.d, SA);
+    else
+        permute_cols_to_left(hmo.layout, hmo.m, hmo.n, selected_idxs.data(), hmo.d, SA);
+
+    //Step 2: Apply the Hadamard transform (invH = H.T = H)
+    int ld = (hmo.left) ? hmo.m : hmo.n;
+
+    T log_sz = std::log2(ld);
+    T log_int_sz, log_final_sz;
+    T log_frac_sz = std::modf(log_sz, &log_int_sz);
+
+    if(log_frac_sz < 1e-3)
+        log_final_sz = log_int_sz;
+    else
+        log_final_sz = log_int_sz + 1;
```

Comment:
This sort of check makes me very uncomfortable. Please handle the problem of computing $\lceil \log_2(\texttt{ld})\rceil$ in a way that's more robust, or provide a proof that this way of doing this is safe.